compiler-core/src/metadata.rs

```diff
@@ -1,10 +1,98 @@
 //! Seriaisation and deserialisation of Gleam compiler metadata into binary files
-//! using the Cap'n Proto schema.
-
-mod module_decoder;
-mod module_encoder;
+//! using the Serde and Bincode
 
 #[cfg(test)]
 mod tests;
 
-pub use self::{module_decoder::ModuleDecoder, module_encoder::ModuleEncoder};
+use crate::type_::{Type, TypeVar};
+use crate::uid::UniqueIdGenerator;
+use crate::{
+    type_::{Module, TypeConstructor},
+    Result,
+};
+use std::cell::RefCell;
+use std::sync::Arc;
+
+pub fn encode(data: &Module) -> Result<Vec<u8>> {
+    let span = tracing::info_span!("metadata");
+    let _enter = span.enter();
+
+    let config = bincode::config::standard();
+    let buffer =
+        bincode::serde::encode_to_vec(data, config).expect("failed to encode metadata (bincode)");
+
+    Ok(buffer)
+}
+
+fn undo_links(id_generator: &UniqueIdGenerator, type_var: TypeVar) -> Type {
+    match type_var {
+        TypeVar::Link { type_ } => decode_type(id_generator, &type_),
+        TypeVar::Generic { .. } => Type::Var {
+            type_: Arc::new(RefCell::new(TypeVar::Generic {
+                id: id_generator.next(),
+            })),
+        },
+        TypeVar::Unbound { .. } => Type::Var {
+            type_: Arc::new(RefCell::new(TypeVar::Unbound {
+                id: id_generator.next(),
+            })),
+        },
+    }
+}
+
+fn decode_type(id_generator: &UniqueIdGenerator, type_: &Type) -> Type {
+    match (*type_).clone() {
+        Type::App {
+            public,
+            module,
+            name,
+            args,
+        } => Type::App {
+            public: public.clone(),
+            module: module.clone(),
+            name: name.clone(),
+            args: args
+                .into_iter()
+                .map(|arg| Arc::new(decode_type(&id_generator, &arg)))
+                .collect::<Vec<Arc<Type>>>(),
+        },
+        Type::Fn { args, retrn } => Type::Fn {
+            args: args
+                .into_iter()
+                .map(|arg| Arc::new(decode_type(&id_generator, &arg)))
+                .collect::<Vec<Arc<Type>>>(),
```

Comment:
You don't need any of the type annotations in this function

Reply:
removed